eip-0017.md

```diff
@@ -0,0 +1,74 @@
+
+- The same whale tried to prevent user's requests (minting/redeeming) from being executed by the assemble service by retuning their funds as soon as funds were broadcasted in the network. This happened also because of the same reasons.
+
+- Moreover, the whale tried to sell SigUSD/SigRSV tokens to users by imitating the bank box. He succeeded to do that and take 2.25% fee for each request which was supposed to go to the SigRSV holders (2%) and UI devs (0.25%).
+
+The above attacks were possible because proxy contracts were not designed to preserve the integrity of the whole dApp infrastructure but only the user's funds.
+
+Generally, proxy contracts should be designed to:
+- prevent dApp developers or any other attacker from taking advantage of user's funds in any manner
+- preserve the integrity of the dApp by preventing attacks like the ones explained in the above examples.
+
+To achieve all of the above, the below contract structure is proposed as an example:
+
+```scala
+{
+  // dApp-specific part ensuring that user will receive what he is paying for
+  val properFundUsage = {
+    val userOut = OUTPUTS(1)
+    userOut.propositionBytes == fromBase64("$userAddress") && // user must be the recipient
+      userOut.tokens(0)._1 == fromBase64("$scTokenId") && // user must receive SigmaUSD
+      userOut.tokens(0)._2 >= $scAmountL && // the amount of SigmaUSD must be at least what user is paying for
+      HEIGHT < $timestampL // this part is always true (timestamp is the unix-timestamp at the time of the request), it will cause compiled address to differ everytime
+  }
+  
+  // ensuring dApp integrity is preserved - any dApp specific condition to ensure designed procedures won't be violated
+  val UIFeeOk = OUTPUTS(2).propositionBytes == fromBase64("$implementor") && OUTPUTS.size == 4 // UI fee must go to UI devs not any random person who assembles the transaction
+  val properBank = OUTPUTS(0).tokens(2)._1 == fromBase64("$bankNFT") // the real bank box of the sigmaUSD protocol must be used so not any random person can behave as the bank box
+  val dAppWorksFine = properFundUsage && UIFeeOk && properBank
+
+  // in any case, whether assembler refuses to execute the request or the request fails for any reason, user must be able to get back his funds
+  val returnFunds = { 
+    val total = INPUTS.fold(0L, {(x:Long, b:Box) => x + b.value}) - $returnFee // only refund transactions's fee must be deducted from user's funds
+    OUTPUTS(0).value >= total && OUTPUTS(0).propositionBytes == fromBase64("$userAddress") // user must receive the appropriate amount
+    && (PK("$assemblerNodeAddr") || HEIGHT > $refundHeightThreshold) && // either dApp-specific node can return user's funds or some time (block) has to be passed first. This is useful for many reasons.
```

Comment:
Wouldn't be better to allow also a user to issue refund transaction in any time (same way as the assembler service could).
So
`&& (PK("$assemblerNodeAddr") || HEIGHT > $refundHeightThreshold)`
would became
`&& ((PK("$assemblerNodeAddr") || PK("$userAddress")) || HEIGHT > $refundHeightThreshold)`
Is there any reason why we don't want the user to have right to do this before the **_$refundHeightThreshold_** threshold has elapsed? :)

Reply:
Explained here: https://github.com/ergoplatform/eips/pull/30#discussion_r640751675